pkg/workflow/step/service_step.go

```diff
@@ -199,3 +201,35 @@ func (s *ServiceStepManager) GenerateTemplates(
 
 	return main, nil, nil
 }
+
+// injectAttributes inject attributes for step.
+func injectAttributes(
+	ctx context.Context,
+	mc model.ClientSet,
+	stepExecution *model.WorkflowStepExecution,
+) (map[string]any, error) {
+	stepAttrs := stepExecution.Attributes
+	// Inject workflow step execution id to request.
+	stepAttrs["workflowStepExecutionID"] = stepExecution.ID.String()
+
+	name, ok := stepExecution.Attributes["name"].(string)
+	if !ok {
+		return nil, errors.New("resource name is not found")
+	}
+
+	res, err := mc.Resources().Query().
+		Where(resource.Name(name)).
+		WithResourceDefinitionMatchingRule().
+		Only(ctx)
+	if err != nil && !model.IsNotFound(err) {
+		return nil, err
+	}
+
+	if res != nil && res.Edges.ResourceDefinitionMatchingRule != nil {
+		stepAttrs["resourceDefinitionMatchingRule"] = map[string]any{
+			"id": res.Edges.ResourceDefinitionMatchingRule.ID.String(),
```

Comment:
How do we use this?

Reply:
Resource validation need `ResourceDefinitionMatchingRuleID` or `templateID` to generate computed attributes, will need add this for resource upgrade API.